File: src/middlewares/logging.ts
import type {
	DescribableInput,
	DescribableOutput,
	LanguageModelMiddleware,
	LanguageModelMiddlewareContext,
	LanguageModelMiddlewareNext,
} from "../describe.ts"

export type LoggingEvent<
	Input extends DescribableInput = DescribableInput,
	Output extends DescribableOutput = DescribableOutput,
> = {
	context: LanguageModelMiddlewareContext<Input, Output>
	result: LanguageModelMiddlewareContext<Input, Output>
	elapsed_ms: number
}

export type LoggingMiddlewareOptions<
	Input extends DescribableInput = DescribableInput,
	Output extends DescribableOutput = DescribableOutput,
> = {log?: (event: LoggingEvent<Input, Output>) => void | Promise<void>}

const default_log = <
	Input extends DescribableInput = DescribableInput,
	Output extends DescribableOutput = DescribableOutput,
>({
	context,
	result,
	elapsed_ms,
}: LoggingEvent<Input, Output>): void => {
	const usage = result.history.at(-1)?.[1].at(-1)?.usage

	console.log(`llm:${context.description.name}`, {
		elapsed_ms,
		prompt_tokens: usage?.prompt_tokens ?? 0,
		completion_tokens: usage?.completion_tokens ?? 0,
		total_tokens: usage?.total_tokens ?? 0,
	})
}

export const logging =
	<
		Input extends DescribableInput = DescribableInput,
		Output extends DescribableOutput = DescribableOutput,
	>(
		options: LoggingMiddlewareOptions<Input, Output> = {},
	): LanguageModelMiddleware<Input, Output> =>
	async (
		context: LanguageModelMiddlewareContext<Input, Output>,
		next: LanguageModelMiddlewareNext<Input, Output>,
	) => {
		const started_at = Date.now()
		const result = await next(context)
		const elapsed_ms = Date.now() - started_at

		const log = options.log ?? default_log<Input, Output>
		await log({context, result, elapsed_ms})

		return result
	}

File: go/middlewares/logging.go
package middlewares

import (
	"log"
	"time"

	"github.com/nmnmcc/toolbox"
)

// LoggingEvent contains information about a logged LLM call.
type LoggingEvent[Input, Output any] struct {
	Context   toolbox.Context[Input, Output]
	Result    toolbox.CompletionContext[Input, Output]
	ElapsedMs int64
}

// LoggingOptions configures logging middleware.
type LoggingOptions[Input, Output any] struct {
	Log func(event LoggingEvent[Input, Output])
}

// DefaultLog provides default logging behavior.
func DefaultLog[Input, Output any](event LoggingEvent[Input, Output]) {
	usage := event.Result.Completion.Usage
	log.Printf("llm:%s elapsed_ms=%d prompt_tokens=%d completion_tokens=%d total_tokens=%d",
		event.Context.Description.Name,
		event.ElapsedMs,
		usage.PromptTokens,
		usage.CompletionTokens,
		usage.TotalTokens,
	)
}

// Logging creates a middleware that logs LLM call metrics.
func Logging[Input, Output any](opts ...LoggingOptions[Input, Output]) toolbox.Middleware[Input, Output] {
	var options LoggingOptions[Input, Output]
	if len(opts) > 0 {
		options = opts[0]
	}
	if options.Log == nil {
		options.Log = DefaultLog[Input, Output]
	}

	return func(ctx toolbox.Context[Input, Output], next toolbox.Next[Input, Output]) (toolbox.CompletionContext[Input, Output], error) {
		start := time.Now()
		result, err := next(ctx)
		elapsed := time.Since(start)

		if err == nil {
			options.Log(LoggingEvent[Input, Output]{
				Context:   ctx,
				Result:    result,
				ElapsedMs: elapsed.Milliseconds(),
			})
		}

		return result, err
	}
}

// LoggingWithLogger creates logging middleware with a custom logger function.
func LoggingWithLogger[Input, Output any](logFn func(event LoggingEvent[Input, Output])) toolbox.Middleware[Input, Output] {
	return Logging(LoggingOptions[Input, Output]{Log: logFn})
}
